In [2]:
# Loading the dataset
import pandas as pd
data = pd.read_csv('weatherHistory.csv')

In [3]:
data.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [4]:
data.shape

(96453, 12)

The dataset consists of 96,453 rows and 12 columns

In [5]:
data.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
count       96453.000000            96453.000000     96453.000000     96453.0   
mean           10.810640              187.509232        10.347325         0.0   
std             6.913571              107.383428         4.192123         0.0   
min             0.000000                0.000000         0.000000         0.0   
25%             5.828200              116.000000         8.339800         0.0   
50%             9.965900              180.000000        10.046400         0.0   
75%            14.135800              290.000000        14.812000         0.0   
max            63.852600              359.000000        16.100000         0.0   

       Pressure (millibars)  
count          96453.000000  
mean            1003.235956  
std              116.969906  
min                0.000000  
25%             1011.900000  
50%             1016.450000  
75%             1021.090000  
max             1046.380000

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


In [7]:
# Checking the datatypes of columns
data.dtypes

Formatted Date               object
Summary                      object
Precip Type                  object
Temperature (C)             float64
Apparent Temperature (C)    float64
Humidity                    float64
Wind Speed (km/h)           float64
Wind Bearing (degrees)      float64
Visibility (km)             float64
Loud Cover                  float64
Pressure (millibars)        float64
Daily Summary                object
dtype: object

In [8]:
# Checking the unique values in the column Precip Type
data['Precip Type'].unique() 

array(['rain', 'snow', nan], dtype=object)

In [9]:
# Checking missing values in the dataset
data.isna().sum()

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
dtype: int64

The column 'Precip Type' has 517 missing values

### Feature selection

In [10]:
# Renaming the columns so it is easier to work with them
data = data.rename(columns = {'Formatted Date': 'Date', 'Precip Type': 'Precip', 'Temperature (C)': 'Temperature', 'Wind Speed (km/h)': 'Wind_Speed', 'Visibility (km)': 'Visibility', 'Pressure (millibars)': 'Pressure'})
data.head()

Date        Summary Precip  Temperature  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy   rain     9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy   rain     9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy   rain     9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy   rain     8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy   rain     8.755556   

   Apparent Temperature (C)  Humidity  Wind_Speed  Wind Bearing (degrees)  \
0                  7.388889      0.89     14.1197                   251.0   
1                  7.227778      0.86     14.2646                   259.0   
2                  9.377778      0.89      3.9284                   204.0   
3                  5.944444      0.83     14.1036                   269.0   
4                  6.977778      0.83     11.0446                   259.0   

   Visibility  Loud Cover  Pressure                      Daily Summary  
0     15.8263         0.0   1015.13  Partly cloudy throughout the day.  
1     15.8263         0.0   1015.63  Partly cloudy throughout the day.  
2     14.9569         0.0   1015.94  Partly cloudy throughout the day.  
3     15.8263         0.0   1016.41  Partly cloudy throughout the day.  
4     15.8263         0.0   1016.51  Partly cloudy throughout the day.

### Handling Missing Values

In [11]:
data['Precip'].isna().sum()

517

In [12]:
# Removing the missing values using dropna and reseting the index
data = data.dropna()

In [13]:
# Checking if the missing values have been removed
data.isnull().sum()

Date                        0
Summary                     0
Precip                      0
Temperature                 0
Apparent Temperature (C)    0
Humidity                    0
Wind_Speed                  0
Wind Bearing (degrees)      0
Visibility                  0
Loud Cover                  0
Pressure                    0
Daily Summary               0
dtype: int64

In [14]:
data['Precip'].unique()

array(['rain', 'snow'], dtype=object)

### Encoding

In [15]:
from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()
data["Precip"]=lb.fit_transform(data["Precip"])

In [16]:
# Checking the counts of unique value in Precip_encode column
data['Precip'].value_counts()

0    85224
1    10712
Name: Precip, dtype: int64

In [17]:
data.dtypes

Date                         object
Summary                      object
Precip                        int32
Temperature                 float64
Apparent Temperature (C)    float64
Humidity                    float64
Wind_Speed                  float64
Wind Bearing (degrees)      float64
Visibility                  float64
Loud Cover                  float64
Pressure                    float64
Daily Summary                object
dtype: object

# Variance Threshold

In [18]:
from sklearn.feature_selection import VarianceThreshold
v=VarianceThreshold(threshold=1)
v.fit(data[["Temperature","Humidity","Wind_Speed","Visibility","Pressure",'Apparent Temperature (C)', 'Wind Bearing (degrees)']])

VarianceThreshold(threshold=1)

In [19]:
del data["Loud Cover"]

In [20]:
v.get_support()

array([ True, False,  True,  True,  True,  True,  True])

### Visualization

In [ ]:
import seaborn as sns
cor = data.corr()
sns.heatmap(data = cor, annot = True)

In [ ]:
d = ['Wind_Speed', 'Visibility', 'Pressure', 'Wind Bearing (degrees)']
dout = ['Wind_Speed', 'Visibility', 'Pressure', 'Wind Bearing (degrees)','Apparent Temperature (C)']
for j in d:
    data.groupby(['Precip']).sum().plot(kind = 'pie', y =j, autopct = '%1.2f%%', explode = [0,0.1])

### Handling Outliers 

In [ ]:
import matplotlib.pyplot as plt
sns.set_style('dark')
data.boxplot(column = ['Temperature'])
plt.show()

In [ ]:
Q1 = data['Temperature'].quantile(.25)
Q3 = data['Temperature'].quantile(.75)
IQR = Q3 - Q1
uw_temp = Q3 + (1.5 * IQR)
lw_temp = Q1 - (1.5 * IQR)
print(uw_temp)
print(lw_temp)

### Log Transformation to check skewness

In [ ]:
import numpy as np
data['log_temperature'] = data['Temperature'].map(lambda i : np.log(i) if i > 0 else 0)
print(data['Temperature'].skew())
print(data['log_temperature'].skew())

### Flooring and Capping

Replacing the values greater than upper whisker with upper whisker which is 40.204 and the values smaller than lower whisker with lower whisker which is -16.756

In [ ]:
data['Temperature'] = np.where(data['Temperature'] < lw_temp , lw_temp, data['Temperature'])
data['Temperature'] = np.where(data['Temperature'] > uw_temp, uw_temp, data['Temperature'])

In [ ]:
data['log_temperature'] = data['Temperature'].map(lambda i : np.log(i) if i > 0 else 0)
print(data['Temperature'].skew())
print(data['log_temperature'].skew())

data.boxplot(column = ['Temperature'])

### Creating a function to handle outliers using flooring and capping method

In [ ]:
def outlier_handle(col):
    print(f'Before handling outliers in {col}')
    data.boxplot(column = [col])
    plt.show()
    Q1 = data[col].quantile(.25)
    Q3 = data[col].quantile(.75)
    IQR = Q3 - Q1
    uw = Q3 + (1.5 * IQR)
    lw = Q1 - (1.5 * IQR)
    print(uw)
    print(lw)
    data[col] = np.where(data[col] < lw , lw, data[col])
    data[col] = np.where(data[col] > uw, uw, data[col])
    print(f'After handling outliers in {col}')
    data.boxplot(column = [col])
    plt.show()
for i in dout:
    outlier_handle(i)

### Model Implementation : Random Forest

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

x=data[['Wind_Speed', 'Visibility', 'Pressure', 'Wind Bearing (degrees)','Apparent Temperature (C)']]
y = data[['Precip']]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)
rf = RandomForestClassifier(n_estimators = 300, max_features = 'sqrt', max_depth = 5, random_state = 18).fit(x_train, y_train)

In [ ]:
y_pred=rf.predict(x_test)

### Checking Accuracy of the model

In [ ]:
rf.score(x_test,y_test)



#### Confusion matrix

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
confusion_matrix(y_test,y_pred)

#### Classification report

In [ ]:
print(classification_report(y_test,y_pred))

# Model implemnetation : Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
logistic = LogisticRegression()
x=data[['Wind_Speed', 'Visibility', 'Pressure', 'Wind Bearing (degrees)','Apparent Temperature (C)']]
y=data["Precip"]
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3)
logistic.fit(x_train, y_train)

In [ ]:
pred = logistic.predict(x_test)

In [ ]:
logistic.score(x_test,y_test)

In [ ]:
confusion_matrix(y_test,pred)

In [ ]:
print(classification_report(y_test,pred))